Monitored commands

erizocosmico opened the pull request:
As discussed in #84, I'll be working on this WIP so you can see if the direction this is going is where you want it to @sdboyer 

The cleaner way to do this, imho, was to wrap each different `vcs.Repo` implementation that is a `vcs.Repo` itself too, so we don't have to reimplement anything. In practice, it has ended up in a lot of copied code from vcs because of not exported code and so on.

I've only implemented the wrapper for git repos so far, so tell me if you want me to keep on this path or do something else entirely.

sdboyer commented:
Sorry, I asked you to put this up, then missed getting to it over the weekend and went on vacation last week :/

Yes, as frustrating as it is to reimplement all this, this seems like the shortest path we can take right now. Carry on!

erizocosmico commented:
I'm so sorry for not replying before. Been kinda busy and I've been snoozing the notification for way too long. Will try to get this done between today and tomorrow.

erizocosmico commented:
@sdboyer Bazaar, Mercurial and Git are done. I saw that SVN is not handled anywhere in the code. Am I missing something or is that intentional? (I implemented the wrapper for SVN as well just in case. If it's intentionally excluded can be removed).

sdboyer commented:
@erizocosmico sweet. I've got this on my list to review soon.

> I saw that SVN is not handled anywhere in the code. Am I missing something or is that intentional?

Not intentional, just...i never got to it, and honestly hadn't thought about it in a while.

> (I implemented the wrapper for SVN as well just in case. If it's intentionally excluded can be removed).

Hah, awesome!

The fact that svn doesn't actually have an out-of-filesystem representation of branches tags makes it pretty awkward to use within our system - there's no multiplexing of versions within a given import path, at all. I do have an idea in mind of how it's supposed to work; I'll compare that against what you've done, here.

sdboyer commented on vcs_repo.go:
```diff
@@ -0,0 +1,300 @@
+		return nil, vcs.NewRemoteError("Unable to retrieve commit information", err, string(out))
+	}
+
+	type Logentry struct {
```
Let's un-export this. (I'm guessing this was just an oversight)

sdboyer commented on vcs_repo.go:
```diff
@@ -0,0 +1,300 @@
+
+			out, err = runFromCwd("git", "clone", r.Remote(), r.LocalPath())
+			if err != nil {
+				return vcs.NewRemoteError("Unable to get repository", err, string(out))
```
I realize this is mostly just copying over from Masterminds/vcs, but let's improve all these errors in two ways:

1. Per standards, don't capitalize the first letter
2. In methods where the same text is used over and over, differentiate the messages. (The repetition might be fine in this method, but there's other places where it potentially obscures what's really going on)

sdboyer commented:
If you rebase on master, the the CircleCI failures should go away.

erizocosmico commented:
Done!

* Rebased (Circle build still broken because it's using my fork and can't find the right one for the example)
* Uncapitalized error messages
* Unexported types used for XML unmarshal
* Added tests for all `vcs.Repo` wrappers

sdboyer commented:
Oh, actually, no, it's failing for a different reason - svn isn't installed (lol). Update `circle.yml` with a command to install it (just add it to the list with bzr), and we can give it another shot.

erizocosmico commented:
Added subversion install, still failing on circle

sdboyer commented:
circle failure still looks spurious. They passed locally; I'm gonna kick circle once more to try for completeness, but I'll merge afterwards either way.

sdboyer commented:
In we go.

sdboyer commented:
If it actually creates problems on my circle, I may pull it back out and figure something out, but...yeah, whatever, we're pretty much there.

Thanks so much for your dogged work on this!
